File: app/yly/algo/todo/cf_81e.py
from app.yly.algo.manage import SolutionBase,View
from common.algo.graph import GraphNode
from typing import Dict,List
from functools import lru_cache
from collections import defaultdict
import bisect
MOD=(10**9)+7
inf = float("inf")
class Node(GraphNode):
    sex=0
    visite=0
class Solution(SolutionBase):
    uri='https://codeforces.com/problemset/problem/81/E'
    def get_cases(self):
        return [
            dict(input='''5
5 2
3 2
5 1
2 1
4 2''',result='''2 2
5 3
2 4''')
        ]
    
    def get_watch(self):
        return [
            View("g")
        ]
    
    def calc(self,u,rt):
        self.f[u][0] = self.f[u][1] = Node(0,0)
        self.g[u]=0
        self.v[u]=1
        for v in self.c[u]:
            if v==rt:
                continue
            self.calc(v,rt)
            self.f[u][0]=self.f[u][0].add(self.f[v][1]) 
            t=self.f[v][0].sub(self.f[v][1]).add(Node(1,self.s[u]^self.s[v]))
            if self.f[u][1].lt(t):
                self.f[u][1]=t
                self.g[u]=v
        self.f[u][1]=self.f[u][1].add(self.f[u][0])

    def get(self,u,i,rt):
        for p in self.c[u]:
            if p==rt:
                continue
            if not i or self.g[u] != p:
                self.get(p,1,rt)
            else:
                self.p.append(Node(u,p))
                self.get(p,0,rt)

    def solve(self,u:Node):
        while not u.visite:
            u=u.visite
            u=u.childs
        r = Node(0,0)
        for _ in range(2):
            self.calc(u,u)
            if r.lt(self.f[u][1]):
                r=self.f[u][1]
                self.p.clear()
                self.get(u,1,u)
            u=self.in_id[u]
        for p in self.p:
            self.q.append(p)
        self.ans=self.ans.add(r)


    def init(self, *args, **kwargs):
        self.n=self.i1()
        self.g = Node().init()
        for i in range(self.n):
            love_id,sex=self.il()
            f,_=self.g.add_edge(i,love_id-1)
            f.sex=sex-1
   

    def exec(self):
        for v in self.g.childs.values():
            if not v.visite:
                self.solve(v)



if __name__=='__main__':
    Solution().run()

File: c/algo/cfs/todo/CF81E.cpp
#include<bits/stdc++.h>
using namespace std;
#define fo(i,a,b) for(int i = a;i <= b;i++)
#define pb push_back
struct node{int x , y;
    bool operator <(node b) {return x == b.x ? y < b.y : x < b.x;} 
    node operator +(node b) const {node ans; ans.x = x + b.x;ans.y = y + b.y;return ans;}
    node operator -(node b) const {node ans; ans.x = x - b.x;ans.y = y - b.y;return ans;}
};
const int N = 1e5 + 5;
int n, in[N] , g[N] , rt , s[N] , v[N];
node f[N][2] , ans;
vector<int> c[N];
vector<node> p , q;
void calc(int u) {
    f[u][1] = f[u][0] = node{0,0}, g[u] = 0, v[u] = 1;
    for(auto P : c[u]) {
        if(P == rt) continue;
        calc(P);
        f[u][0] = f[u][0] + f[P][1];
        node t = f[P][0] - f[P][1] + node{1 ,s[u] ^ s[P]};
        if(f[u][1] < t) f[u][1] = t , g[u] = P;
    }
    f[u][1] = f[u][1] + f[u][0];
}
void get(int u , int i) {
    for(auto P : c[u])  { 
        if(P == rt) continue;
        if(!i || g[u] != P) get(P , 1); else p.pb(node{u,P}) , get(P , 0);
    }
}
void solve(int u) {
    for(;!v[u];u = in[u]) v[u] = 1; // 找环上的点
    node r = node{0,0};
    fo(i,1,2) {
        rt = u; // 断边
        calc(u);
        if(r < f[u][1]) r = f[u][1] , p.clear() , get(u,1);
        u = in[u]; // 换边
    }
    for(auto P : p) q.pb(P);
    ans = ans + r;
}
int main() {
    scanf("%d" , &n);
    fo(i,1,n) scanf("%d %d",in + i , s + i) , s[i]--,c[in[i]].pb(i);
    fo(i,1,n) if(!v[i]) solve(i);
    printf("%d %d\n" , ans.x , ans.y);
    for(auto P:q) printf("%d %d\n" , P.x , P.y);
    return 0;
}